packages/jest-resolve-dependencies/src/index.ts

```diff
@@ -51,19 +52,47 @@ class DependencyResolver {
       if (this._resolver.isCoreModule(dependency)) {
         return acc;
       }
+
       let resolvedDependency;
+      let resolvedMockDependency;
       try {
         resolvedDependency = this._resolver.resolveModule(
           file,
           dependency,
           options,
         );
       } catch (e) {
-        resolvedDependency = this._resolver.getMockModule(file, dependency);
+        // TODO: This snippet might not be necessary, should be investigated later.
+        const mockDependency = this._resolver.getMockModule(file, dependency);
+        mockDependency && acc.push(mockDependency);
+        return acc;
+      }
+
+      if (!resolvedDependency) {
+        return acc;
       }
 
-      if (resolvedDependency) {
-        acc.push(resolvedDependency);
+      acc.push(resolvedDependency);
+
+      // If we resolve a dependency, then look for a mock dependency
+      // of the same name in that dependency's directory.
+      resolvedMockDependency = this._resolver.getMockModule(
+        path.dirname(resolvedDependency),
```

Comment:
Is there a reason why it's not `file, dependency` like above? Would be nice if we could just `try` to resolve, ignoring failure, and then look for a mock in a single call regardless of whether the resolution failed, to simplify the logic here

Reply:
`getMockModule(file, dependency)` actually does not work as expected.

Taking another look at it though, what does work with the tests that I added is this:

```typescript
  resolvedMockDependency = this._resolver.getMockModule(
    dependency
    path.basename(dependency),
  );
```

Which is definitely better than what I had before.

Unfortunately, this will fail for the case where the dependency is in a different directory from the dependent, and there's still the existing issue of mocks of the same name from different directories causing issues.

If this isn't good enough, I'm thinking we'll have to either introduce a new method into `jest-resolve` which mirrors the functionality resolveModule, or expand the functionality of `resolveModule` to handle this case, as well as to avoid some of the hackyness of the rest of my solution.

I do feel as though I'm slightly out of my depth here since it seems like some fairly important changes need to happen to pick up those cases, so if anyone has any guidance that would be appreciated.

Reply:
I don't understand, sorry. I would expect that `getMockModule` does no more than look for a `__mocks__/file.js` next to the `file.js`. Why does it matter where the dependent is, it's not being passed to `getMockModule`?